### ISSUE TITLE
Reserve customized modules in torch.compile's dynamo tracer

### ISSUE THREAD

**Issue by trajepl** [pytorch/pytorch#149062]
## Description
Hi PyTorch team,

I’ve encountered an issue with torch.compile when working with customized modules. Specifically, torch.compile tends to step into customized modules and decompose them into built-in functions and modules. This leads to the loss of the original module information, which makes it difficult to replace the modules or analyze the computation graph in torch.fx.

torch.fx.symbolic_trace has a similar issue, but it provides the `is_leaf_module` interface ([docs](https://pytorch.org/docs/stable/fx.html#torch.fx.Tracer.is_leaf_module)) that allows developers to treat certain modules as leaf nodes and prevent them from being decomposed. This makes it possible to retain the module identity within the traced graph.

However, in my case, torch.fx.symbolic_trace doesn't work well because my model contains many custom operators that are not compatible with symbolic tracing.

## Feature Request
Would it be possible to introduce a similar `is_leaf_module` interface for torch.compile? This would give developers more control over how torch.compile treats custom modules, helping preserve module structure and improving the flexibility of graph transformations.

## Example
Here's a minimal example to illustrate the issue:
```python
import torch
import torch.nn as nn

class CustomModule(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return nn.functional.linear(x, torch.randn(10, 10))

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.custom = CustomModule()

    def forward(self, x):
        return self.custom(x)

model = Model()
compiled_model = torch.compile(model)
input = torch.randn(1, 10)
output = compiled_model(input)
print(output)
```

### Current Behavior
Now, the gx graph in torch.compile looks like:

```
opcode         name    target                                                    args           kwargs
-------------  ------  --------------------------------------------------------  -------------  --------
placeholder    l_x_    L_x_                                                      ()             {}
call_function  randn   <built-in method randn of type object at 0x7f87b3f94380>  (10, 10)       {}
call_function  linear  <built-in function linear>                                (l_x_, randn)  {}
output         output  output                                                    ((linear,),)   {}
```

### Expected Behavior
If torch.compile had an `is_leaf_module`-like interface, we could define a custom function to treat CustomModule as a leaf. For example
```python
def my_leaf_module_check_func(module):
    return isinstance(module, CustomModule)
    
compiled_model = torch.compile(model, options={'is_leaf_module': my_leaf_module_check_func})
```
The expected graph would preserve the CustomModule identity:
```
opcode       name              target            args                    kwargs
-----------  ----------------  ----------------  ----------------------  --------
placeholder  l_x_              L_x_              ()                      {}
call_module  l__self___custom  L__self___custom  (l_x_,)                 {}
output       output            output            ((l__self___custom,),)  {}
```

**Comment by shunting314** [comment]
@anijain2305 do we already have such thing in dynamo to step over a module?

**Comment by wschin** [comment]
A more fundamental ask is probably "keeping module structure but replacing all callables with aten ops".

**Comment by ice-tong** [comment]
Related https://discuss.pytorch.org/t/how-to-customize-leaf-node-in-dynamo-trace/195718
